AL-943 reorganization of documentation and addition of user introduction (#604)
spacetelescope/gwcs: docs/gwcs/user_introduction.rst

.. _user_intro:

Using Generalized World Coordinate System Models
================================================

Brief Introduction and Outline
------------------------------

`GWCS (Generalized World Coordinate System) <https://github.com/spacetelescope/gwcs>`__
package is intended as a generalization of the FITS WCS standards, which basically provides

> a transformation from detector array coordinates to sky, spectral, time or any other physical

Reviewer: I'm going to suggest a strict difference between "pixel" and "array" coordinates as used by APE 14, where "array" is in numpy (C) order. See [the ndcube docs](https://docs.sunpy.org/projects/ndcube/en/stable/introduction.html#important-concepts-array-pixel-and-world-axes) for how we explain this to our users.

```suggestion
a transformation from detector pixel coordinates to sky, spectral, time or any other physical
```

Author: This seems reasonable and this change has been made. However...

coordinates, and the converse: from the physical coordinates to the corresponding detector
array coordinates. While simple in principle, it can be a complex subject. The main focus
of this documentation is first the user, who only needs to know how to use GWCS objects,
but doesn't care about the complexities or the internals of how they work, nor the justification
for this new standard. Those interested in those internals or the motiviations for GWCS
will find this material in later sections.

The documentation is organized into the following sections:

* User Perspective: How to use GWCS models, usually packaged with the data.
* Saving and reading GWCS models.
* Developer Perspective: How to create and modify GWCS models.
* Motivation: Why GWCS was developed and how it is a large improvement over FITS WCS.

User Perspective
----------------

As mentioned, the focus here is on using GWCS models provided by others, most likely, models
that are packaged with the data they apply to.

.. _pixel-conventions-and-definitions:

> Terminology and Conventions

Author: Looking at your link, it only addresses the axis order, but not the 0 vs 1 convention. Do you see that as independent of the pixel/array distinction?

> Terminology and Conventions

Reviewer: Yeah, I think it's independent. In that both pixel and arrays can be indexed from either 0 or 1. APE 14 indexes both from 0 for example.

The greatest source of confusion
will likely be the the presumed order of pixel coordinates. There are two main 
conventions for ordering pixel coordinates. In the context of
2-dimensional imaging data/arrays, one can either think of the pixel coordinates
as traditional Cartesian coordinates (which we call ``x`` and ``y`` here), which
are usually given with the horizontal coordinate (``x``) first, and the vertical
coordinate (``y``) second, meaning that pixel coordinates would be given as
``(x, y)``. Alternatively, one can give the coordinates by first giving the row
in the data, then the column, i.e. ``(row, column)``. While the former is a more
common convention when e.g. plotting (think for example of the Matplotlib
``scatter(x, y)`` method), the latter is the convention used when accessing
values from e.g. Numpy arrays that represent images (``image[row, column]``).

The GWCS object assumes Cartesian order ``(x, y)``, however it should be mentioned
that there is an WCS shared API for both GWCS and FITS WCS that can use
either ordering. The details regarding the shared API will be found later.

.. _installing-gwcs:

Installing GWCS
...............

> To install the latest release::
> 
>     pip install gwcs

Reviewer 2: ```suggestion
To install the latest release:

.. code-block:: shell

    pip install gwcs
```

To install as a conda package from `conda-forge <https://github.com/conda-forge/gwcs-feedstock>`__:

>     conda install -c conda-forge gwcs

Reviewer 2: ```suggestion

.. code-block:: shell

    conda install -c conda-forge gwcs
```

To install the latest development version from source (not generally recommended one 

> needs a very new feature or bug fix)::
> 
>      pip install git+https://github.com/spacetelescope/gwcs.git

Reviewer 2: ```suggestion
needs a very new feature or bug fix):

.. code-block:: shell

     pip install git+https://github.com/spacetelescope/gwcs.git
```

> If the clone has already been done:
> 
>      cd gwcs
>      pip install .

Reviewer 2: ```suggestion
If you wish to install directly from source with the ability to edit the source code:

.. code-block:: shell

  git clone https://github.com/spacetelescope/gwcs.git
  cd gwcs
  pip install -e gwcs
```

of this documentation is first the user, who only needs to know how to use GWCS objects,
but doesn't care about the complexities or the internals of how they work, nor the justification
for this new standard. Those interested in those internals or the motiviations for GWCS 
will find this material in later sections.

The documentation is organized into the following sections:

* User Perspective: How to use GWCS models, usually packaged with the data.
* Saving and reading GWCS models.
* Developer Perspective: How to create and modify GWCS models.
* Motivation: Why GWCS was developed and how it is a large improvement over FITS WCS.

User Perspective
----------------

As mentioned, the focus here is on using GWCS models provided by others, most likely, models
that are packaged with the data they apply to.

Terminology and Conventions
...........................

To use WCS models (of any kind, FITS or GWCS) it is important to understand the terminology 
and conventions, particularly with regard to pixel coordinates. For example, are the pixel
coordinates 0 or 1 based? That is, is the first pixel identified as 0 or 1? GWCS assumes 0 

> (whereas FITS assumes 1; do not confuse the two!). However, like FITS, integer pixel values 

Author: Done.

are presumed to refer to the center of the pixel. So for GWCS, the boundaries of the first 
pixel are at -0.5 and +0.5 for all dimensions (typically 2).

The greatest source of confusion
will likely be the the presumed order of pixel coordinates. There are two main 
conventions for ordering pixel coordinates. In the context of
2-dimensional imaging data/arrays, one can either think of the pixel coordinates
as traditional Cartesian coordinates (which we call ``x`` and ``y`` here), which
are usually given with the horizontal coordinate (``x``) first, and the vertical
coordinate (``y``) second, meaning that pixel coordinates would be given as
``(x, y)``. Alternatively, one can give the coordinates by first giving the row
in the data, then the column, i.e. ``(row, column)``. While the former is a more
common convention when e.g. plotting (think for example of the Matplotlib
``scatter(x, y)`` method), the latter is the convention used when accessing
values from e.g. Numpy arrays that represent images (``image[row, column]``).

The GWCS object assumes Cartesian order ``(x, y)``, however it should be mentioned
that there is an Astropy common API for both GWCS and FITS WCS that can use

> either ordering. The details regarding the common API will be found later. It may

Author: Done.

> or may not be relevant one's needs.

Author: Done.

ASDF extension with ASDF when installed. In other words, when GWCS is installed,
ASDF understands how to save and recover GWCS objects. 
The structure of an ASDF file can be considered as a dictionary (technically,
including lists as well) where the "keys" are attributes of the nested dictionaries.
If a value of any of these attributes is an GWCS object, it will be converted
into a form that ASDF knows how to save in the file, and upon reading, the
corresponding information will be turned back into a GWCS object in Python 
(Note that ASDF is language neutral, and implementations in other languages
should be able to construct equivalent objects for GWCS in that languages though
none yet exist).

The following example illustrates how easily this can be done

Continuing with the example of the previous spectrograph GWCS case.

.. doctest-skip::

 >>> af2 = asdf.AsdfFile() # Create a new ASDF object
 >>> af2.tree = {'wcs': wcs} # Only saving gwcs object in this example
 >>> af2.write_to('my_spectral_wcs.asdf')
 >>> af3 = asdf.open('my_spectral_wcs.asdf') # read it back into memory
 >>> wcs2 = af3.tree['wcs']
 >>> wcs2 == wcs # Confirm it is the same as the one originally stored.

>  True

Author: Can you be more specific?

And that is all there is to it

The only format that GWCS supports at this time is ASDF.

JWST currently embeds GWCS information in FITS files as an ASDF FITS extension.

Using the Universal WCS API
...........................

Astropy has developed a common API for WCS libraries that permits both FITS
WCS and GWCS libraries to be accessed using the same methods and attributes,
allowing application code to work with both kinds of objects. It was developed
under APE 14 (Astropy Proposal for Enhancement)

The general documentation for this API can be found
`here <https://github.com/astropy/astropy-APEs/blob/2bed2c015ef59459f5ef44adc12ea6d72eea7e6f/APE14.rst>`__

This API consists of low-level and high-level methods. This section will only 
deal with the high-level API, and only briefly. The main functionality is
represented by these methods:

* pixel_to_world() For example, wcs.pixel_to_world(x, y) instead of wcs(x, y)

> * world_to_pixel() wcs.world_to_pixel(ra, dec) instead of wcs.invert(ra, dec)

Reviewer 3: We can add this in a follow up PR

> * world_to_pixel() wcs.world_to_pixel(ra, dec) instead of wcs.invert(ra, dec)

Author: This almost deserves its own page. So you are suggesting removing all this for the time being?
